### PR TITLE
OPENEUROPA-3387: Fix translated media rendering for paragraphs and Event.

### PR DESCRIPTION
**Pull request by 22Alexandra**

### CODE REVIEW

**Review by imanoleguskiza — CHANGES_REQUESTED** [review]
The fixes are good but we need to add tests on oe_theme. I sugges creating kernel tests for these.

**Comment by 22Alexandra** [comment]
> The fixes are good but we need to add tests on oe_theme. I sugges creating kernel tests for these.

I added kernel test for the paragraphs but as discussed it was tricky testing the media rendering in kernel for event and in behat the file was not uploaded.

File: oe_theme.theme [hunk 1]
```diff
@@ -962,6 +962,9 @@ function oe_theme_preprocess_paragraph__oe_text_feature_media(array &$variables)
   /** @var \Drupal\paragraphs\Entity\Paragraph $paragraph */
   $paragraph = $variables['paragraph'];
 
+  // Retrieve the correct paragraph translation.
+  $paragraph = \Drupal::service('entity.repository')->getTranslationFromContext($paragraph, $paragraph->language()->getId());
```
Comment (upchuk) [inline oe_theme.theme]: Why are you passing a language here? It defaults to the current content language but you are passing the original paragraph language so it will do...nothing?

File: oe_theme.theme [hunk 2]
```diff
@@ -1403,6 +1410,8 @@ function oe_theme_preprocess_paragraph__oe_timeline(array &$variables): void {
 function oe_theme_preprocess_paragraph__oe_banner(array &$variables): void {
   /** @var \Drupal\paragraphs\Entity\Paragraph $paragraph */
   $paragraph = $variables['paragraph'];
+  // Retrieve the correct paragraph translation.
+  $paragraph = \Drupal::service('entity.repository')->getTranslationFromContext($paragraph, $paragraph->language()->getId());
```
Comment (upchuk) [inline oe_theme.theme]: Same here on the paragraph translation.

File: tests/Kernel/ContentEventRenderTest.php [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+use Drupal\file\Entity\File;
+use Drupal\Tests\rdf_entity\Traits\RdfDatabaseConnectionTrait;
+use Drupal\user\Entity\Role;
+use Drupal\user\RoleInterface;
+use Symfony\Component\DomCrawler\Crawler;
+
+/**
+ * Tests that our Event content type render.
+ *
+ * @todo: Extend this test with more scenarios.
```
Comment (upchuk) [inline tests/Kernel/ContentEventRenderTest.php]: What scenarios is it missing? This todo is quite meaningless.

File: tests/Kernel/ContentEventRenderTest.php [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+ */
+class ContentEventRenderTest extends AbstractKernelTestBase {
```
Comment (upchuk) [inline tests/Kernel/ContentEventRenderTest.php]: Cant we extend from `ContentRenderTest ` ?

File: tests/Kernel/Paragraphs/ContentRowTest.php [hunk 5]
```diff
@@ -152,9 +151,6 @@ public function testRendering(): void {
    * Tests that side menu title translations are rendered correctly.
    */
   public function testSideMenuTranslation(): void {
-    // Create "French" language.
-    ConfigurableLanguage::createFromLangcode('fr')->save();
```
Comment (upchuk) [inline tests/Kernel/Paragraphs/ContentRowTest.php:156]: Why did you remove? Where is this created?

**Review by upchuk — CHANGES_REQUESTED** [review]

File: tests/Kernel/Paragraphs/ContentRowTest.php (hunk 5 above)
Comment (22Alexandra) [reply tests/Kernel/Paragraphs/ContentRowTest.php:156]: The languages are created in ParagraphsTestBase.

File: tests/Kernel/ContentEventRenderTest.php (hunk 4 above)
Comment (nagyad) [reply tests/Kernel/ContentEventRenderTest.php]: Because here we don't need news, pages, publication and policy content types to be installed, but many other for event content type related modules only. Therefore the parent::setup method can't be executed but the one in one more level down, the AbstractKernelTest.

File: oe_theme.theme (hunk 1 above)
Comment (nagyad) [reply oe_theme.theme]: You can see now the tests are failing. I noticed during debugging without that it get's reset to the English version so the English get's rendered instead of the Bulgarian.

File: tests/Kernel/ContentEventRenderTest.php (hunk 4 above)
Comment (ademarco) [reply tests/Kernel/ContentEventRenderTest.php]: Can't we turn this into a browser test? I've tried it locally and it sets up in 85 seconds. 

image

All you need to do is to replace the initial part of this test with the following:

```php
class ContentEventRenderTest extends BrowserTestBase {

  /**
   * {@inheritdoc}
   */
  public static $modules = [
    'config',
    'system',
    'oe_theme_helper',
    'path',
    'oe_theme_content_event',
    'content_translation',
  ];

  /**
   * {@inheritdoc}
   */
  protected function setUp() {
    parent::setUp();

    // Enable and set OpenEuropa Theme as default.
    $this->container->get('theme_installer')->install(['oe_theme']);

    // Make event node and image media translatable.
    $this->container->get('content_translation.manager')->setEnabled('node', 'oe_event', TRUE);
    $this->container->get('content_translation.manager')->setEnabled('media', 'image', TRUE);
    $this->container->get('router.builder')->rebuild();
  }
```

Comment (nagyad) [reply tests/Kernel/ContentEventRenderTest.php]: Yes that was my initial try, but to debug this is a pure nightmare. Your example is missing couple things. The whole operation to have the event node created with files and translations, takes me 2.40mins to run. To make this test work I need to run it several times, I can't afford 20 minutes of waiting if I run this test 10 times. To debug it's double the time easily, and it's counter-productive. The only way I could imagine to go for such robust tests is with a local class that skips the setup after the first run and restores a database dump from the point when the setup finished and before anything ran. I tried, really and I lost hours...

Comment (nagyad) [reply tests/Kernel/ContentEventRenderTest.php]: Also, for single scenario it could be okay, but as soon as you need to check 4-5 content types and different test scenarios the time quickly increases from 84 seconds to 500 seconds impacting the project test time on drone.

File: oe_theme.theme [hunk 6]
```diff
@@ -975,6 +975,10 @@ function oe_theme_preprocess_paragraph__oe_text_feature_media(array &$variables)
     // The media entity is not available anymore, bail out.
     return;
   }
+
+  // Retrieve the correct media translation.
+  $media = \Drupal::service('entity.repository')->getTranslationFromContext($media, $paragraph->language()->getId());
```
Comment (ademarco) [inline oe_theme.theme:980]: We don't need this right, as it will get it from the context?
```
$paragraph->language()->getId()
```

File: oe_theme.theme [hunk 7]
```diff
@@ -1452,6 +1456,9 @@ function oe_theme_preprocess_paragraph__oe_banner(array &$variables): void {
     return;
   }
 
+  // Retrieve the correct translation to display.
+  $media = \Drupal::service('entity.repository')->getTranslationFromContext($media, $paragraph->language()->getId());
```
Comment (ademarco) [inline oe_theme.theme:1460]: Same as above.

File: tests/Functional/ContentEventRenderTest.php [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+      'uri' => 'public://example_1.jpeg',
+    ]);
+    $en_image->save();
+
+    // Create image media that we will use for the Bulgarian translation.
+    $this->container->get('file_system')->copy(drupal_get_path('theme', 'oe_theme') . '/tests/fixtures/placeholder.png', 'public://placeholder.png');
+    $bg_image = File::create([
+      'uri' => 'public://placeholder.png',
+    ]);
+    $bg_image->save();
```
Comment (ademarco) [inline tests/Functional/ContentEventRenderTest.php]: Let's use the same as we use below, for consistency and it's also more appropriate (with the _en,  _bg etc.):

```php
    $en_file = file_save_data(file_get_contents(drupal_get_path('theme', 'oe_theme') . '/tests/fixtures/example_1.jpeg'), 'public://example_1_en.jpeg');
    $en_file->setPermanent();
    $en_file->save();

    // Create Bulgarian file.
    $bg_file = file_save_data(file_get_contents(drupal_get_path('theme', 'oe_theme') . '/tests/fixtures/example_1.jpeg'), 'public://example_1_bg.jpeg');
    $bg_file->setPermanent();
    $bg_file->save();
```

File: tests/Functional/ContentEventRenderTest.php [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+    $file_urls = [
+      'en' => $en_image->createFileUrl(),
+      'bg' => $bg_image->createFileUrl(),
+    ];
+
+    foreach ($node->getTranslationLanguages() as $node_langcode => $node_language) {
+      $node = $this->container->get('entity.repository')->getTranslationFromContext($node, $node_langcode);
+      $this->drupalGet($node->toUrl());
+      $this->assertSession()->elementExists('css', 'img[src*="' . $file_urls[$node_langcode] . '"]');
```
Comment (ademarco) [inline tests/Functional/ContentEventRenderTest.php]: Let's restrict the scope of the selector in this test to the actual element where the image is supposed to be displayed, this is a bit too broad: `img[src*="' . $file_urls[$node_langcode] . '"]`

**Review by ademarco — CHANGES_REQUESTED** [review]

File: oe_theme.theme (hunk 6 above)
Comment (nagyad) [reply oe_theme.theme:980]: If I remove these, tests are failing as it loads the English version of the file.

**Review by ademarco — APPROVED** [review]

**Review by imanoleguskiza — APPROVED** [review]

**Review by upchuk — APPROVED** [review]
